### PR TITLE
WQ Factory Custom Worker Binary

### PR DESCRIPTION
**Pull request by Nekel-Seyew**
This is the addition of the feature request in issue #1604. Tested on the Disc machines to work.

### CODE REVIEW

File: batch_job/src/work_queue_factory.c [hunk 1]
```diff
@@ -99,6 +99,7 @@ static char *batch_submit_options = NULL;
 
 static char *wrapper_command = 0;
 static char *wrapper_input = 0;
+static char *worker_prog = 0;
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: worker_command

File: batch_job/src/work_queue_factory.c [hunk 2]
```diff
@@ -316,14 +317,33 @@ static void set_worker_resources_options( struct batch_queue *queue )
 	resource_args = xxstrdup(buffer_tostring(&b));
 	buffer_free(&b);
 }
-
+static char* fetch_end_of_path(char* path){
+	size_t len= strlen(path);
+	size_t slash_pos = 0;
+	size_t i=0;
+	for(i=len-1; i > -1; i-- ){
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: man strrchr

File: batch_job/src/work_queue_factory.c [hunk 3]
```diff
@@ -316,14 +317,33 @@ static void set_worker_resources_options( struct batch_queue *queue )
...
+		if(path[i] == '/'){
+			slash_pos = i;
+			break;
+		}
+	}
+	char* new_str = malloc(sizeof(char)*(len-slash_pos)+1);
+	for(i=0; i< (len-slash_pos); i++){
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: man strncpy

File: batch_job/src/work_queue_factory.c [hunk 4]
```diff
@@ -316,14 +317,33 @@ static void set_worker_resources_options( struct batch_queue *queue )
 	resource_args = xxstrdup(buffer_tostring(&b));
 	buffer_free(&b);
 }
-
+static char* fetch_end_of_path(char* path){
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: We usually use fetch for transfers over the network. 

Also, see:  dttools/src/path.h:path_basename

File: batch_job/src/work_queue_factory.c [hunk 5]
```diff
@@ -316,14 +317,33 @@ static void set_worker_resources_options( struct batch_queue *queue )
...
+		new_str[i] = path[slash_pos + i];
+	}
+	new_str[len-slash_pos] = '\0';
+	return new_str;
+}
 static int submit_worker( struct batch_queue *queue )
 {
 	char *cmd;
+	char *tmp = fetch_end_of_path(worker_prog);
+	char *worker = (worker_prog != NULL) ? tmp : string_format("./work_queue_worker");
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: Better:
const char *worker = worker_command ? path_basename(worker_command) : "./work_queue_worker";

File: batch_job/src/work_queue_factory.c [hunk 6]
```diff
@@ -355,7 +376,7 @@ static int submit_worker( struct batch_queue *queue )
 		cmd = newcmd;
 	}
 
-	char *files = string_format("work_queue_worker");
+	char *files = (worker_prog != NULL) ? string_format("%s",worker_prog) : string_format("work_queue_worker");
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: As above.

File: batch_job/src/work_queue_factory.c [hunk 7]
```diff
@@ -370,6 +391,7 @@ static int submit_worker( struct batch_queue *queue )
 	}
 
 	debug(D_WQ,"submitting worker: %s",cmd);
+	free(worker);
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: If followed, the above suggestions would make this free unnecessary.

File: batch_job/src/work_queue_factory.c [hunk 8]
```diff
@@ -899,6 +937,7 @@ static const struct option long_options[] = {
 	{"condor-requirements", required_argument, 0, LONG_OPT_CONDOR_REQUIREMENTS},
 	{"wrapper",required_argument, 0, LONG_OPT_WRAPPER},
 	{"wrapper-input",required_argument, 0, LONG_OPT_WRAPPER_INPUT},
+	{"custom-worker", required_argument, 0, LONG_OPT_CUSTOM_WORKER},
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: worker-binary

File: batch_job/src/work_queue_factory.c (hunk 5 above)
Comment (matz-e) [inline batch_job/src/work_queue_factory.c]: There's some inconsistency here?
```
const char *worker = worker_command ? path_basename(worker_command) : "work_queue_worker";
```
In one part of the command below `./%s` is userd, in the other `%s`. Why not change them both to `./%s` and drop the `./` here in front of `work_queue_worker`?

**Comment by dthain** [comment]
@Nekel-Seyew you are approach the problem from the wrong end.  We don't want to change how the worker is invoked, just the location of the original binary.  Look for the copy command at the end of `main`, that's all you need to change.

**Comment by Nekel-Seyew** [comment]
Made the changes

**Pull request merged**